File: android/app/src/b3SandboxProof/java/uk/eugnel/ks2spelling/B3ProofObservationPlugin.java
package uk.eugnel.ks2spelling;

import android.content.Intent;
import android.system.Os;
import android.system.OsConstants;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.io.FileDescriptor;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import org.json.JSONObject;

@CapacitorPlugin(name = "B3ProofObservation")
public final class B3ProofObservationPlugin extends Plugin {
    public static final String COMMAND_EXTRA =
        "uk.eugnel.ks2spelling.B3_PROOF_COMMAND_V1";
    private static final String OBSERVATION_FILENAME =
        "b3-proof-observation-v1.json";
    private static final int MAXIMUM_BYTES = 64 * 1024;
    private static final Set<String> OBSERVATION_KEYS = new HashSet<>(Arrays.asList(
        "schemaVersion", "platform", "buildAuthoritySha256", "captureId",
        "installationId", "sequence", "previousObservationSha256", "scenarioIndex",
        "scenario", "phase", "nextActionCode", "completedTransitions",
        "proofProjection", "observedAt", "observationSha256"
    ));

    @PluginMethod public void getLaunchCommand(PluginCall call) {
        try {
            require(BuildConfig.B3_SANDBOX_PROOF && call.getData().length() == 0);
            Intent intent = getActivity().getIntent();
            String value = intent == null ? null : intent.getStringExtra(COMMAND_EXTRA);
            JSObject result = new JSObject();
            if (value == null) {
                result.put("commandJson", JSONObject.NULL);
            } else {
                byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
                require(bytes.length > 0 && bytes.length <= MAXIMUM_BYTES);
                result.put("commandJson", value);
            }
            call.resolve(result);
        } catch (Exception error) {
            reject(call);
        }
    }

    @PluginMethod public void publishObservation(PluginCall call) {
        File temporary = null;
        try {
            require(BuildConfig.B3_SANDBOX_PROOF);
            requireExactKeys(call.getData(), setOf("canonicalJson"));
            String canonicalJson = call.getString("canonicalJson");
            require(canonicalJson != null);
            byte[] bytes = canonicalJson.getBytes(StandardCharsets.UTF_8);
            require(bytes.length > 0 && bytes.length <= MAXIMUM_BYTES);
            validateClosedObservation(canonicalJson);

            File root = getContext().getExternalFilesDir(null);
            require(root != null && root.isDirectory());
            require(!Files.isSymbolicLink(root.toPath()));
            File destination = new File(root, OBSERVATION_FILENAME);
            assertSafeTarget(destination, false);

            temporary = File.createTempFile("b3-proof-observation-", ".tmp", root);
            require(temporary.getParentFile().getCanonicalFile().equals(root.getCanonicalFile()));
            require(!Files.isSymbolicLink(temporary.toPath()) && temporary.isFile());
            try (FileOutputStream output = new FileOutputStream(temporary, false)) {
                output.write(bytes);
                output.getFD().sync();
            }
            Files.move(
                temporary.toPath(),
                destination.toPath(),
                StandardCopyOption.ATOMIC_MOVE,
                StandardCopyOption.REPLACE_EXISTING
            );
            temporary = null;
            syncDirectory(root);
            assertSafeTarget(destination, true);
            JSObject result = new JSObject();
            result.put("written", true);
            call.resolve(result);
        } catch (Exception error) {
            if (temporary != null) temporary.delete();
            reject(call);
        }
    }

    private static void validateClosedObservation(String canonicalJson) throws Exception {
        JSONObject object = new JSONObject(canonicalJson);
        Set<String> keys = new HashSet<>();
        Iterator<String> iterator = object.keys();
        while (iterator.hasNext()) keys.add(iterator.next());
        require(keys.equals(OBSERVATION_KEYS) && object.length() == OBSERVATION_KEYS.size());
    }

    private static void assertSafeTarget(File target, boolean mustExist) throws Exception {
        require(OBSERVATION_FILENAME.equals(target.getName()));
        if (!target.exists()) {
            require(!mustExist);
            return;
        }
        require(!Files.isSymbolicLink(target.toPath()));
        require(target.isFile() && target.length() > 0 && target.length() <= MAXIMUM_BYTES);
    }

    private static void syncDirectory(File directory) throws Exception {
        FileDescriptor descriptor = Os.open(
            directory.getAbsolutePath(),
            OsConstants.O_RDONLY,
            0
        );
        try {
            Os.fsync(descriptor);
        } finally {
            Os.close(descriptor);
        }
    }

    private static void requireExactKeys(JSONObject object, Set<String> expected) throws Exception {
        Set<String> keys = new HashSet<>();
        Iterator<String> iterator = object.keys();
        while (iterator.hasNext()) keys.add(iterator.next());
        require(keys.equals(expected) && object.length() == expected.size());
    }

    private static Set<String> setOf(String... values) {
        return new HashSet<>(Arrays.asList(values));
    }

    private static void require(boolean condition) throws Exception {
        if (!condition) throw new Exception("rejected");
    }

    private static void reject(PluginCall call) {
        call.reject("Proof observation rejected.", "B3_PROOF_OBSERVATION_REJECTED");
    }
}

File: ios/App/App/B3ProofObservationPlugin.swift
import Capacitor
import Foundation

#if B3_SANDBOX_PROOF
@objc(B3ProofObservationPlugin)
public final class B3ProofObservationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "B3ProofObservationPlugin"
    public let jsName = "B3ProofObservation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getLaunchCommand", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "publishObservation", returnType: CAPPluginReturnPromise)
    ]

    private static let commandArgument = "--b3-proof-command-v1"
    private static let observationFilename = "b3-proof-observation-v1.json"
    private static let maximumBytes = 64 * 1024
    private static let observationKeys: Set<String> = [
        "schemaVersion", "platform", "buildAuthoritySha256", "captureId",
        "installationId", "sequence", "previousObservationSha256", "scenarioIndex",
        "scenario", "phase", "nextActionCode", "completedTransitions",
        "proofProjection", "observedAt", "observationSha256"
    ]

    @objc public func getLaunchCommand(_ call: CAPPluginCall) {
        guard call.options.isEmpty else {
            reject(call)
            return
        }
        let arguments = ProcessInfo.processInfo.arguments
        let positions = arguments.indices.filter {
            arguments[$0] == Self.commandArgument
        }
        guard positions.count <= 1 else {
            reject(call)
            return
        }
        guard let position = positions.first else {
            call.resolve(["commandJson": NSNull()])
            return
        }
        let valueIndex = arguments.index(after: position)
        guard valueIndex < arguments.endIndex,
              arguments[valueIndex] != Self.commandArgument,
              let bytes = arguments[valueIndex].data(using: .utf8),
              !bytes.isEmpty,
              bytes.count <= Self.maximumBytes else {
            reject(call)
            return
        }
        call.resolve(["commandJson": arguments[valueIndex]])
    }

    @objc public func publishObservation(_ call: CAPPluginCall) {
        do {
            try requireKeys(call, exactly: ["canonicalJson"])
            guard let canonicalJson = call.getString("canonicalJson"),
                  let bytes = canonicalJson.data(using: .utf8),
                  !bytes.isEmpty,
                  bytes.count <= Self.maximumBytes else {
                throw ProofTransportError.rejected
            }
            try validateClosedObservation(bytes)
            let destination = try observationURL()
            try assertSafeTarget(destination)
            try bytes.write(
                to: destination,
                options: [.atomic, .completeFileProtection]
            )
            try assertSafeTarget(destination, mustExist: true)
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
            var mutableDestination = destination
            try mutableDestination.setResourceValues(values)
            call.resolve(["written": true])
        } catch {
            reject(call)
        }
    }

    private func observationURL() throws -> URL {
        let support = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let values = try support.resourceValues(
            forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
        )
        guard values.isDirectory == true, values.isSymbolicLink != true else {
            throw ProofTransportError.rejected
        }
        var backupValues = URLResourceValues()
        backupValues.isExcludedFromBackup = true
        var mutableSupport = support
        try mutableSupport.setResourceValues(backupValues)
        return support.appendingPathComponent(Self.observationFilename, isDirectory: false)
    }

    private func assertSafeTarget(_ target: URL, mustExist: Bool = false) throws {
        guard target.lastPathComponent == Self.observationFilename else {
            throw ProofTransportError.rejected
        }
        if !FileManager.default.fileExists(atPath: target.path) {
            if mustExist { throw ProofTransportError.rejected }
            return
        }
        let values = try target.resourceValues(
            forKeys: [.isRegularFileKey, .isSymbolicLinkKey, .fileSizeKey]
        )
        guard values.isRegularFile == true,
              values.isSymbolicLink != true,
              let size = values.fileSize,
              size > 0,
              size <= Self.maximumBytes else {
            throw ProofTransportError.rejected
        }
    }

    private func validateClosedObservation(_ bytes: Data) throws {
        let value = try JSONSerialization.jsonObject(with: bytes)
        guard let object = value as? [String: Any],
              Set(object.keys) == Self.observationKeys,
              object.keys.count == Self.observationKeys.count else {
            throw ProofTransportError.rejected
        }
    }

    private func requireKeys(_ call: CAPPluginCall, exactly expected: Set<String>) throws {
        let keys = Set(call.options.keys.compactMap { $0 as? String })
        guard keys == expected, call.options.keys.count == keys.count else {
            throw ProofTransportError.rejected
        }
    }

    private func reject(_ call: CAPPluginCall) {
        call.reject("Proof observation rejected.", ProofTransportError.rejected.safeCode)
    }
}

private enum ProofTransportError: Error {
    case rejected

    var safeCode: String { "B3_PROOF_OBSERVATION_REJECTED" }
}
#endif
